### ISSUE TITLE
allow `map(f, ::String)` for `f` not producing chars

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaLang/julia#24562]
Currently `map(f, "Hello, world.")` only works if `f` maps characters to characters. I think that's a good behavior, but it would be good if when `f` produces a non-character, it would produce an array instead. This is somewhat similar to how the element type of a `map` result depends on the actual values returned by the mapping function.

**Comment by JeffBezanson** [comment]
I don't think this is necessary. Typically map produces a container of the same kind, and it's easy enough to pass a generator to `collect` or another collection constructor to get other kinds of result values.

**Comment by martinholters** [comment]
We already have:
```julia
julia> typeof(map(string, falses(3)))
Array{String,1}

julia> typeof(map(identity, falses(3)))
BitArray{1}

julia> typeof(map(identity, zeros(Bool, 3)))
Array{Bool,1}
```
Here, the output container type depends on the generated element type and the input container type, so applying the same idea to strings as input containers sounds reasonable.

**Comment by StefanKarpinski** [comment]
It's certainly not necessary since we've been getting by without it, but it would be a pleasant feature and would have a pleasant DWIMiness about it, much like our other map methods. Can be added at any point since it's currently an error for the transform not to return characters.

**Comment by TotalVerb** [comment]
This can't possibly be type stable since `map(f, "")` is always ""::String.

**Comment by martinholters** [comment]
Of course, we would have to change that, too. (Replace the `3` above with `0` and you get the same results.) But that would be breaking... We could relatively easily add a `depwarn` if string is empty and `!(Char <: return_type(f, Tuple{Char})` or similar now.

**Comment by TotalVerb** [comment]
Unfortunately `Array` and `String`, unlike `Array` and `BitArray`, do not generally have the same interface. So we can't return an empty `Vector{Union{}}` when potentially the caller expects a string, and we can't safely return an empty `String` when potentially the caller expects a `Vector{Int}`.

**Comment by StefanKarpinski** [comment]
Presumably the call knows what function it's applying and therefore what `map` will return. Moreover, the compiler knows that too since it specializes on function arguments.

**Comment by StefanKarpinski** [comment]
Anyway, this is not really worth discussion at this point; I'm just leaving this here as an idea for some point in the future.

**Comment by TotalVerb** [comment]
The problem is that that is not safe in general:

```
julia> function f(x::Char)
           value = 0
           for c in '\0':x
               value = c
           end
           value
       end
f (generic function with 1 method)

julia> Core.Inference.return_type(f, Tuple{Char})
Union{Char, Int64}
```

This should have `f("") === ""`, but relying on inference to figure that out is not possible.

**Comment by adienes** [comment]
another idea: `map(f, s::String)` could mean `join(map(f, collect(s)), "")`

**Comment by StefanKarpinski** [comment]
I think we could actually just broaden the return type: if the function returns only chars then we return a string, if it returns anything else then we return a vector. We can use return type internally to optimize this. I think it's fine since it won't break any existing code.

**Comment by jariji** [comment]
My view of `map` (https://github.com/JuliaLang/julia/issues/57075) is that it should not be defined on types where it fails to always preserve `keys(_)`.  `String` is one of those types (https://github.com/JuliaLang/julia/issues/57071#issuecomment-2596931308).

**Comment by StefanKarpinski** [comment]
Yeah, pretty much everyone agrees that in an ideal world the special case for map on strings should be removed. If there was no collateral damage to changing that, we would. But we don't live in that world and we can't break compatibility. So the question is what can we do to make this less annoying when one wants to map a function that isn't `Char --> Char` over a string.
